lib/pigeon/socket_tracker.ex

```diff
@@ -0,0 +1,39 @@
+defmodule Pigeon.SocketTracker do
+  @moduledoc false
+  require Logger
+
+  @ets_peer_table Pigeon.Application.ets_peer_table()
+
+  @doc """
+  Check in the ETS table if the given socket (according to its peername) is a duplicate. If
+  it is and if `allow_duplicates` is false, then the socket is added to the table and
+  `{:error, :duplicate}` is returned. Otherwise the connect response is returned as-is.
+  """
+  @spec check_duplicate(term(), boolean()) ::
+          {:ok, any()} | {:error, :duplicate} | any()
+  def check_duplicate(connect_response, allow_duplicates)
+
+  def check_duplicate({:ok, socket}, false) do
+    with {:ok, peername} <- Pigeon.Shared.peername(socket),
+         dbg(peername),
+         true <- :ets.insert_new(@ets_peer_table, {peername}) do
+      IO.puts("***** unique")
+      {:ok, socket}
+    else
+      false ->
+        IO.puts("***** duplicate")
+        Process.unlink(socket)
+        Pigeon.Http2.Client.default().close(socket)
```

Comment:
This line is the one that seems to cause the problem. If I comment this one out, then everything ALMOST works as expected, except that if I call `Troubleshooting.PushNotifications.fetch_push_endpoint_connection_info()` then I can see that there are open sockets left lying around to the duplicates that presumably no one will ever use.

Reply:
I'm assuming the `Process.unlink` is here to prevent the dispatcher worker from getting the `handle_info({:closed, _}, ...` method called from the process closing -- but what if that isn't working quite as expected?

What if we let the dispatcher worker handle reconnecting like it already does when the socket is closed, and instead just call `Pigeon.Http2.Client.default().close()` here. I think the recursive call on duplicate would need to be removed from `lib/pigeon/fcm.ex` for that to work correctly, but could be something to try.

Reply:
The problem is that we're still in the startup process, so if one of the adapters returns an error, the application crashes and everything just shuts down.